libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,294 @@
+                Location test_loc2 = test_loc;
+                test_loc2.offset_bearing(bearing_test2, distance2);
+
+                // calculate minimum margin to fence and obstacles for this scenario
+                float margin2 = calc_avoidance_margin(test_loc, test_loc2);
+                if (margin2 > _margin_max) {
+                    // all good, now project in the chosen direction by the full distance
+                    destination_new = current_loc;
+                    destination_new.offset_bearing(bearing_test, distance_to_dest);
+                    _current_lookahead = MIN(_lookahead, _current_lookahead * 1.1f);
+                    return (i != 0 || j != 0);
+                }
+            }
+        }
+    }
+
+    float chosen_bearing;
+    if (have_best_bearing) {
+        // none of the directions tested were OK for 2-step checks. Choose the direction
+        // that was best for the first step
+        chosen_bearing = best_bearing;
+        _current_lookahead = MIN(_lookahead, _current_lookahead * 1.05f);
+    } else {
+        // none of the possible paths had a positive margin. Choose
+        // the one with the highest margin
+        chosen_bearing = best_margin_bearing;
+        _current_lookahead = MAX(_lookahead * 0.5f, _current_lookahead * 0.9f);
+    }
+
+    // calculate new target based on best effort
+    destination_new = current_loc;
+    destination_new.offset_bearing(chosen_bearing, distance_to_dest);
+
+    return true;
+}
+
+// calculate minimum distance between a path and any obstacle
+float AP_OABendyRuler::calc_avoidance_margin(const Location &start, const Location &end)
+{
+    float circular_fence_margin;
+    if (!calc_margin_from_circular_fence(start, end, circular_fence_margin)) {
+        circular_fence_margin = FLT_MAX;
+    }
+
+    float polygon_fence_margin;
+    if (!calc_margin_from_polygon_fence(start, end, polygon_fence_margin)) {
+        polygon_fence_margin = FLT_MAX;
+    }
+
+    float proximity_margin;
+    if (!calc_margin_from_proximity_sensors(start, end, proximity_margin)) {
+        proximity_margin = FLT_MAX;
+    }
+
+    // return smallest margin from any obstacle
+    return MIN(MIN(circular_fence_margin, polygon_fence_margin), proximity_margin);
+}
+
+// calculate minimum distance between a path and the circular fence (centered on home)
+// on success returns true and updates margin
+bool AP_OABendyRuler::calc_margin_from_circular_fence(const Location &start, const Location &end, float &margin)
+{
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
+        return false;
+    }
+    if ((fence_ptr->get_enabled_fences() & AC_FENCE_TYPE_CIRCLE) == 0) {
+        return false;
+    }
+
+    // calculate start and end point's distance from home
+    const Location ahrs_home = AP::ahrs().get_home();
```

Comment:
```suggestion
    const Location &ahrs_home = AP::ahrs().get_home();
```

Reply:
txs for that.  By the way, we have a similar mistake in GCS_Common.cpp's GCS_MAVLINK::send_home_position.  I might create a small PR for this fix.